src/miral/minimal_window_manager.cpp

```diff
@@ -277,6 +298,13 @@ auto miral::MinimalWindowManager::confirm_inherited_move(WindowInfo const& windo
 void miral::MinimalWindowManager::advise_new_window(miral::WindowInfo const& window_info)
 {
     self->application_selector.advise_new_window(window_info);
+
+    // If focus stealing prevention is on, swap the old focused window (now in
+    // the back) with the new window in the front.
+    // If it's a legitimate window, it'll be focused and raised via
+    // xdg-activation.
+    if ((self->focus_stealing == FocusStealing::prevent) && tools.active_window())
+        tools.swap_tree_order(tools.active_window(), window_info.window());
```

Comment:
I'm not convinced that this is exactly right.

1. Consider there being a fullscreen window with focus when a notification window is created. I don't expect the notification to be activated (as it shouldn't get focus), but it should be above the fullscreen window. Trying to push it "below" seems wrong (even if it ultimately fails because of an implementation detail: the windows are not in the same layer);
2. Consider an application that creates a new window, e.g. a menu. This should go on top without needing activation. Here the call fails with `log_error("Unable to swap tree order due to overlap");` because BasicWindowManager checks for an overlap between the trees. Spamming the log with errors is a bad idea;
3. In addition to the `SurfaceStack`, `application_selector` is maintaining a stack of windows. I don't see this swap being notified (it will still have the new window on top). Having inconsistencies between the two stacks will surely lead to problems.

Reply:
On 1., what is "a notification window"? How would we distinguish it from any other surface?

My original bug report about stealing did mention notifications, because that's how it affected me. But _notifications_ should never be toplevels, and we can't be expected to treat them differently to any other toplevel.

Reply:
> we can't be expected to treat them differently to any other toplevel

This code is handling ALL windows, not just toplevels

Reply:
Looking at window types:
```
    mir_window_type_normal,       /**< AKA "regular"                       */
    mir_window_type_utility,      /**< AKA "floating"                      */
    mir_window_type_dialog,
    mir_window_type_gloss,
    mir_window_type_freestyle,
    mir_window_type_menu,
    mir_window_type_inputmethod,  /**< AKA "OSK" or handwriting etc.       */
    mir_window_type_satellite,    /**< AKA "toolbox"/"toolbar"             */
    mir_window_type_tip,          /**< AKA "tooltip"                       */
    mir_window_type_decoration,
```

Maybe we can only swap `mir_window_type_{normal,utility}`?

Reply:
For 3., the first thing that comes to mind is overloading `advise_new_window` to sync the two stacks together

Reply:
I proposes that the windows that should be swapped are `mir_window_type_normal`, in the "application" layer, and parentless; anything else can be left on top

Reply:
Seems like some wayland applications use `mir_window_type_freestyle` as well.

Reply:
OK: in the "application" layer, and parentless

Reply:
I relaxed the condition and tried around with different applications and things seem to work as expected.